lec05-dplyr.Rmd

```diff
@@ -1,4 +1,510 @@
+argument `fill=0`.
+
+```{r}
+surveys_gw %>%
+  spread(genus, mean_weight, fill = 0) %>%
+  head
+```
+
+We can now do things like plot the weight of *Baiomys* against *Chaetodipus* or
+examine their correlation.
+
+```{r}
+surveys_gw %>%
+  spread(genus, mean_weight, fill = 0) %>%
+  cor(use = "pairwise.complete")
+```
+
+### Wide to long with `gather`
+
+What if we had the opposite problem, and wanted to go from a wide to long
+format? For that, we use `gather` to sweep up a set of columns into one
+key-value pair. We give it the arguments of a new key and value column name, and
+then we specify which columns we either want or do not want gathered up. So, to
+go backwards from `surveys_gw_wide`, and exclude `plot_id` from the gathering,
+we would do the following:
+
+```{r}
+surveys_gw_long <- surveys_gw_wide %>%
+  gather(genus, mean_weight, -plot_id)
+
+head(surveys_gw_long)
+```
+
+Note that now the `NA` genera are included in the long format. Going from wide
+to long to wide can be a useful way to balance out a dataset so every replicate
+has the same composition.
+
+We could also have used a specification for what columns to include. This can be
+useful if you have a large number of identifying columns, and it's easier to
+specify what to gather than what to leave alone. And if the columns are in a
+row, we don't even need to list them all out - just use the `:` operator!
+
+```{r}
+surveys_gw_wide %>%
+  gather(genus, mean_weight, Baiomys:Spermophilus) %>%
+  head()
+```
+
+># ### Challenge
+>
+> 1. Make a wide data frame with `year` as columns, `plot_id` as rows, and the
+>   values are the number of genera per plot. You will need to summarize before
+>   reshaping, and use the function `n_distinct` to get the number of unique
+>   types of a genera. It's a powerful function! See `?n_distinct` for more.
+>
+> 2. Now take that data frame, and make it long again, so each row is a unique
+>    `plot_id` - `year` combination.
+>
+> 3. The `surveys` data set is not truly wide or long because there are
+>    two columns of measurement - `hindfoot_length` and `weight`.  This makes it
+>    difficult to do things like look at the relationship between mean values of
+>    each measurement per year in different plot types. Let's walk through a
+>    common solution for this type of problem. First, use `gather` to create a
+>    truly long dataset where we have a key column called `measurement` and a
+>    `value` column that takes on the value of either `hindfoot_length` or
+>    `weight`. Hint: You'll need to specify which columns are being gathered.
+>
+> 4. With this new truly long data set, calculate the average of each
+>    `measurement` in each `year` for each different `plot_type`. Then
+>    `spread` them into a wide data set with a column for `hindfoot_length` and
+>    `weight`. Hint: Remember, you only need to specify the key and value
+>    columns for `spread`.
+
+<!---
+  ```{r, echo=FALSE, purl=FALSE}
```

Comment:
What's the `purl` for?

Reply:
https://youtu.be/MAMD_XpybQY

Reply:
Hahaha, they really stick to the theme ey?

Reply:
Ha! I didn't realize that was a real thing, haha wow!